cunumeric/module.py

```diff
@@ -6227,3 +6227,693 @@ def bincount(
+        range(min_dim_index + num_axes, ndim)
+    )
+    reshaped = tuple(
+        [
+            collapsed_shape if k == min_dim_index else arr_shuffled.shape[k]
+            for k in redimed
+        ]
+    )
+
+    arr_reshaped = arr_shuffled.reshape(reshaped)
+    return (min_dim_index, arr_reshaped)
+
+
+# args:
+#
+# arr:      [in] source nd-array on which quantiles are calculated;
+#                preccondition: assumed sorted!
+# q_arr:    [in] quantile input values nd-array;
+# axis:     [in] axis along which quantiles are calculated;
+# method:   [in] func(q, n) returning (gamma, j),
+#                where = array1D.size;
+# keepdims: [in] boolean flag specifying whether collapsed axis
+#                should be kept as dim=1;
+# to_dtype: [in] dtype to convert the result to;
+# qs_all:   [in/out] result pass through or created (returned)
+#
+def quantile_impl(
+    arr,
+    q_arr,
+    axis,
+    axes_set,
+    original_shape,
+    method,
+    keepdims,
+    to_dtype,
+    qs_all,
+):
+
+    ndims = len(arr.shape)
+
+    if axis is None:
+        n = arr.size
+
+        if keepdims:
+            remaining_shape = tuple(ones(len(axes_set), dtype=int))
+        else:
+            remaining_shape = ()  # only `q_arr` dictates shape;
+        # quantile applied to `arr` seen as 1D;
+    else:
+        n = arr.shape[axis]
+
+        # arr.shape -{axis}; if keepdims use 1 for arr.shape[axis]:
+        # (can be empty [])
+        #
+        if keepdims:
+            remaining_shape = [
+                1 if k in axes_set else original_shape[k]
+                for k in range(0, len(original_shape))
+            ]
+        else:
+            remaining_shape = [
+                arr.shape[k] for k in range(0, ndims) if k != axis
+            ]
+
+    # compose qarr.shape with arr.shape:
+    #
+    # result.shape = (q_arr.shape, arr.shape -{axis}):
+    #
+    qresult_shape = (*q_arr.shape, *remaining_shape)
+
+    # fails with cunumeric:
+    # (numpy works)
+    #
+    # cunumeric BUG: TypeError: a bytes-like object is required, not 'ndarray';
+    # TODO: open nvbug
+    #
+    # qs_all = ndarray(shape = qresult_shape,
+    #                     buffer = array(result_1D),
+    #                     dtype = to_dtype)
+    #
+    # construct result NdArray, non-flattening approach:
+    #
+    if qs_all is None:
+        qs_all = zeros(qresult_shape, dtype=to_dtype)
+    else:
+        # implicit conversion from to_dtype to qs_all.dtype assumed
+        #
+        if qs_all.shape != qresult_shape:
+            raise ValueError("wrong shape on output array")
+
+    for index, q in np.ndenumerate(q_arr):
+        (gamma, j) = method(q, n)
+        (left_pos, right_pos) = (j, j + 1)
+
+        # (N-1) dimensional ndarray of left, right
+        # neighbor values:
+        #
+        # non-flattening approach:
+        #
+        # extract values at index=left_pos;
+        arr_1D_lvals = arr.take(left_pos, axis)
```

Comment:
Even the `left_pos` apparently can be out-of-bounds. E.g. in this case `j` comes out to `4`
```
print(cn.quantile(cn.arange(4), 1.0, method="weibull"))
```
and thus this crashes when `take` is called.

Reply:
Fixed.